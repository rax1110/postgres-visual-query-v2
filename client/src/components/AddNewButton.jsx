import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import React from 'react';
import * as PropTypes from 'prop-types';
import { Button } from 'reactstrap';

export const AddNewButton = ({ id, size, onClick }) => (
  <Button
    className=""
    outline
    color="info"
    size={size}
    id={id}
    onClick={onClick}
  >
    <FontAwesomeIcon icon="plus" />
  </Button>
);

AddNewButton.propTypes = {
  id: PropTypes.string,
  size: PropTypes.string,
  onClick: PropTypes.func,
};
