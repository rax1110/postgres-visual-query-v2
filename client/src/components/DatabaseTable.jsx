import React from 'react';
import { Button, Tooltip } from 'reactstrap';
import { connect } from 'react-redux';
import _ from 'lodash';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as PropTypes from 'prop-types';
import { addColumn, addTable, removeTable } from '../actions/queryActions';
import { withToggle } from '../hocs/withToggle';

export const DatabaseTable = ({ data, checked, id, addTableProp, toggle, toggleStatus }) => {
  const handleOnClick = () => {
    addTableProp(data);
  };

  let tableTypeColor = 'primary';

  if (_.isEqual(data.table_type, 'VIEW')) {
    tableTypeColor = 'info';
  } else if (_.isEqual(data.table_type, 'FOREIGN')) {
    tableTypeColor = 'secondary';
  } else if (_.isEqual(data.table_type, 'MATERIALIZED VIEW')) {
    tableTypeColor = 'light';
  }

  const btnSelected = checked ? 'success' : tableTypeColor;

  const modifiers = {
    preventOverflow: {
      enabled: false,
    },
    hide: {
      enabled: false,
    },
  };

  return (
    <div className="w-100 pr-1">
      <Button
        size="sm"
        color={btnSelected}
        id={id}
        className="btn-block my-1 pt-0 text-left"
        onClick={handleOnClick}
      >
        <small
          color={tableTypeColor}
          className="text-truncate align-self-start"
        >
          <span className="mr-1 px-0">
            <FontAwesomeIcon icon="th-large" />
          </span>
          {data.table_type}
        </small>
        <div className="text-truncate">
          {` ${data.table_name}`}
        </div>
      </Button>
      <Tooltip
        placement="right"
        isOpen={toggleStatus}
        target={id}
        toggle={toggle}
        modifiers={modifiers}
        delay={{ show: 200, hide: 0 }}
        className=""
      >
        {data.table_name}
      </Tooltip>
    </div>
  );
};

DatabaseTable.propTypes = {
  data: PropTypes.shape({ table_type: PropTypes.string, table_name: PropTypes.string }),
  checked: PropTypes.bool,
  id: PropTypes.string,
  addTableProp: PropTypes.func,
  toggle: PropTypes.func,
  toggleStatus: PropTypes.bool,
};

const mapDispatchToProps = {
  addColumn,
  addTableProp: data => addTable(data),
  removeTable,
};

export default withToggle(connect(null, mapDispatchToProps)(DatabaseTable));
