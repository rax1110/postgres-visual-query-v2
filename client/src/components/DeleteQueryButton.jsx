import React from 'react';
import { Button } from 'reactstrap';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as PropTypes from 'prop-types';
import { deleteQuery, setActiveQuery } from '../actions/queryActions';
import { deleteQueries, removeMainFromQueries } from '../actions/queriesActions';

export const DeleteQueryButton = (props) => {
  const handleOnClick = () => {
    props.deleteQuery();

    const mainQuery = props.queries.find(query => query.id === 0);

    if (mainQuery) {
      props.removeMainFromQueries();
      props.setActiveQuery(mainQuery);
    } else {
      props.deleteQueries();
    }
  };

  return (
    <Button color="danger" className="mr-2" onClick={handleOnClick}>
      <FontAwesomeIcon icon="trash" />
    </Button>
  );
};

DeleteQueryButton.propTypes = {
  deleteQuery: PropTypes.func,
  deleteQueries: PropTypes.func,
  queries: PropTypes.arrayOf(PropTypes.shape({})),
  removeMainFromQueries: PropTypes.func,
  setActiveQuery: PropTypes.func,
};

const mapStateToProps = store => ({
  queries: store.queries,
});

const mapDispatchToProps = {
  deleteQuery,
  deleteQueries,
  removeMainFromQueries,
  setActiveQuery,
};

export default connect(mapStateToProps, mapDispatchToProps)(DeleteQueryButton);
