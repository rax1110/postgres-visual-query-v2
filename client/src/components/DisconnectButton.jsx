import React from 'react';
import { connect } from 'react-redux';
import { Button } from 'reactstrap';
import { Redirect } from 'react-router-dom';
import * as PropTypes from 'prop-types';
import { disconnect } from '../actions/hostActions';
import { translations } from '../utils/translations';

export const DisconnectButton = (props) => {
  const handleOnClick = () => {
    props.disconnect();
  };

  if (!props.connected) {
    return <Redirect to="/" />;
  }

  return (
    <Button size="lg" className="btn-block my-2 px-2" color="danger" onClick={handleOnClick}>
      {translations[props.language.code].sideBar.disconnectB}
    </Button>
  );
};

DisconnectButton.propTypes = {
  disconnect: PropTypes.func,
  connected: PropTypes.bool,
  language: PropTypes.shape({ code: PropTypes.string }),
};

const mapStateToProps = store => ({
  connected: store.host.connected,
  language: store.settings.language,
});

const mapDispatchToProps = {
  disconnect,
};

export default connect(mapStateToProps, mapDispatchToProps)(DisconnectButton);
