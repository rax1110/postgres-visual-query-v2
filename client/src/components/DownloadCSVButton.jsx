import React from 'react';
import { connect } from 'react-redux';
import { CSVLink } from 'react-csv';
import _ from 'lodash';
import { Button } from 'reactstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as PropTypes from 'prop-types';

export const DownloadCSVButton = (props) => {
  const getHeaders = (result) => {
    if (_.isNull(result)) {
      return [];
    }

    const headers = [];

    result.fields.forEach((field) => {
      const header = {
        label: field.name,
        key: field.name,
      };

      headers.push(header);
    });

    return headers;
  };

  const getData = (result) => {
    if (_.isNull(result)) {
      return [];
    }
    return result.rows;
  };

  const disabled = _.isNull(props.result);

  return (
    <CSVLink
      className="mr-2"
      data={getData(props.result)}
      headers={getHeaders(props.result)}
      filename="result.csv"
    >
      <Button disabled={disabled}>
        <FontAwesomeIcon icon="download" />
        <div className="d-inline"> CSV</div>
      </Button>
    </CSVLink>
  );
};

DownloadCSVButton.propTypes = {
  result: PropTypes.shape({}),
};

const mapStateToProps = store => ({
  result: store.query.result,
});

export default connect(mapStateToProps)(DownloadCSVButton);
