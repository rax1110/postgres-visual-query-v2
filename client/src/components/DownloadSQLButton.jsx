import React from 'react';
import { connect } from 'react-redux';
import { Button } from 'reactstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import * as PropTypes from 'prop-types';

export const DownloadSQLButton = (props) => {
  const downloadContent = (name, content) => {
    if (navigator.msSaveBlob) {
      const blobObject = new Blob([content], { type: 'text/plain' });

      window.navigator.msSaveOrOpenBlob(blobObject, name);
    } else {
      const aTag = document.createElement('a');
      const file = new Blob([content], { type: 'text/plain' });

      aTag.href = URL.createObjectURL(file);
      aTag.download = name;

      document.body.appendChild(aTag);
      aTag.click();
    }
  };

  return (
    <Button
      className="mr-2"
      onClick={() => downloadContent('select_query.sql', props.sql)}
    >
      <FontAwesomeIcon icon="download" />
      <div className="d-inline"> SQL</div>
    </Button>
  );
};

DownloadSQLButton.propTypes = {
  sql: PropTypes.string,
};

const mapStateToProps = store => ({
  sql: store.query.sql,
});

export default connect(mapStateToProps)(DownloadSQLButton);
