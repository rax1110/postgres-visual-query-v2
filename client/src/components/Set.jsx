import React, { useEffect } from 'react';
import { Draggable } from 'react-beautiful-dnd';
import {
  Button,
  Card,
  CardBody,
  Container,
  CustomInput,
  Form,
  FormGroup,
  Row,
  UncontrolledTooltip,
} from 'reactstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import _ from 'lodash';
import * as PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { removeSet, updateSet } from '../actions/queryActions';
import { translations } from '../utils/translations';
import { getCorrectQueryName } from '../utils/getCorrectQueryName';

export const Set = (props) => {
  useEffect(() => {
    if (props.set.subqueryId) {
      let set = _.cloneDeep(props.set);

      const subquery = props.queries.find(query => query.id === props.set.subqueryId);
      const subquerySql = subquery ? subquery.sql : '';

      set = {
        ...set,
        subquerySql,
      };

      props.updateSet(set);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [props.set.subqueryId]);

  const selectedQuery = props.queries.find(query => query.id === props.set.subqueryId);
  const selectedQueryId = selectedQuery ? selectedQuery.id : 0;

  const handleTypeChange = (e) => {
    e.preventDefault();

    let set = _.cloneDeep(props.set);

    set = {
      ...set,
      type: e.target.value,
    };

    props.updateSet(set);
  };

  const handleQueryChange = (e) => {
    e.preventDefault();

    let set = _.cloneDeep(props.set);

    const subqueryId = +e.target.value;
    const subquerySql = subqueryId ? props.queries.find(query => query.id === subqueryId).sql : '';

    set = {
      ...set,
      subqueryId,
      subquerySql,
    };

    props.updateSet(set);
  };

  const handleRemove = () => {
    props.removeSet(props.set);
  };

  return (
    <div className="my-2">
      <Draggable
        draggableId={props.id}
        index={props.index}
      >
        {provided => (
          <Card
            {...provided.draggableProps}
            {...provided.dragHandleProps}
            innerRef={provided.innerRef}
          >
            <CardBody className="py-2 px-0">
              <Form>
                <Container fluid>
                  <Row>
                    <div className="col-auto d-flex">
                      <FontAwesomeIcon className="align-self-center" icon="sort" />
                    </div>
                    <div className="col-10 px-0">
                      <Row form className="align-items-center">
                        <div className="col-auto pb-1">
                          <span style={{
                            color: props.set.color,
                            fontWeight: 'bold',
                            fontSize: 'large',
                          }}
                          >
                            {'{}'}
                          </span>
                        </div>
                        <div className="col-auto">
                          {props.index === 0
                            ? getCorrectQueryName(props.language, props.queryName, props.queryId)
                            : translations[props.language.code].queryBuilder.setResult}
                        </div>
                        <div className="col-auto">
                          <FormGroup className="m-0">
                            <CustomInput
                              bsSize="sm"
                              type="select"
                              id={`${props.id}-set-type`}
                              onChange={handleTypeChange}
                              value={props.set.type}
                              key={`set-type-${props.id}-query-${props.queryId}`}
                            >
                              <option value="union">UNION</option>
                              <option value="unionall">UNION ALL</option>
                              <option value="intersect">INTERSECT</option>
                              <option value="except">EXCEPT</option>
                            </CustomInput>
                            <UncontrolledTooltip
                              placement="top"
                              delay={{ show: 500, hide: 0 }}
                              target={`${props.id}-set-type`}
                            >
                              {translations[props.language.code].tooltips.setType}
                            </UncontrolledTooltip>
                          </FormGroup>
                        </div>
                        <div className="col-5">
                          <FormGroup className="m-0">
                            <CustomInput
                              bsSize="sm"
                              type="select"
                              name="select_query"
                              id={`set-query-${props.id}`}
                              key={`set-${props.id}-query-${props.queryId}`}
                              value={selectedQueryId}
                              onChange={handleQueryChange}
                            >
                              <option value="">
                                {translations[props.language.code]
                                  .queryBuilder.setQuery}
                              </option>
                              {props.queries.map(query => (
                                <option
                                  key={`set-${props.id}-query-${query.id}`}
                                  value={query.id}
                                >
                                  {getCorrectQueryName(props.language, query.queryName, query.id)}
                                </option>
                              ))}
                            </CustomInput>
                          </FormGroup>
                        </div>
                      </Row>
                    </div>
                    <div className="col-1 d-flex ml-auto pr-2 justify-content-end">
                      <FormGroup className="align-self-center m-0">
                        <Button
                          size="sm"
                          color="danger"
                          onClick={handleRemove}
                          id={`${props.id}-remove-set`}
                        >
                          <FontAwesomeIcon icon="times" />
                        </Button>
                      </FormGroup>
                    </div>
                  </Row>
                </Container>
              </Form>
            </CardBody>
          </Card>
        )}
      </Draggable>
    </div>
  );
};

Set.propTypes = {
  id: PropTypes.string,
  language: PropTypes.shape({ code: PropTypes.string }),
  set: PropTypes.shape({
    id: PropTypes.number,
    color: PropTypes.string,
    type: PropTypes.string,
    subqueryId: PropTypes.number,
  }),
  updateSet: PropTypes.func,
  removeSet: PropTypes.func,
  index: PropTypes.number,
  queries: PropTypes.arrayOf(PropTypes.shape({})),
  queryName: PropTypes.string,
  queryId: PropTypes.number,
};

const mapStateToProps = store => ({
  language: store.settings.language,
  queries: store.queries.filter(query => query.id !== 0)
    .sort((query1, query2) => query1.id - query2.id),
});

const mapDispatchToProps = {
  updateSet,
  removeSet,
};

export default connect(mapStateToProps, mapDispatchToProps)(Set);
