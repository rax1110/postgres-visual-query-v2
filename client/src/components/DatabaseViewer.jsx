import React from 'react';
import { connect } from 'react-redux';
import { Scrollbars } from 'react-custom-scrollbars';
import _ from 'lodash';
import * as PropTypes from 'prop-types';
import DatabaseTable from './DatabaseTable';
import { filterTable } from '../utils/filterTable';

export const DatabaseViewer = (props) => {
  const constructData = (table) => {
    const data = {
      table_schema: table.table_schema,
      table_name: table.table_name,
      table_type: table.table_type,
      table_alias: '',
    };

    let constraints = JSON.parse(JSON.stringify(props.constraints));

    constraints = constraints.filter(constraint => constraint.table_schema === data.table_schema
      && constraint.table_name === data.table_name);

    let columns = JSON.parse(JSON.stringify(props.columns));

    columns = columns.filter(column => column.table_name === data.table_name
      && column.table_schema === data.table_schema).map((column) => {
      const col = column;

      col.constraints = constraints.filter(
        constraint => _.includes(constraint.column_name, column.column_name),
      );

      delete col.table_name;
      delete col.table_schema;
      return col;
    });

    data.columns = columns;

    return data;
  };

  return (
    <div className="flex-fill">
      <Scrollbars className="d-flex" autoHide>
        <div className="mt-1 pr-2">
          {props.tables.map((table, index) => {
            const checked = props.queryTable.some(
              queryTable => _.isEqual(table.table_name, queryTable.table_name)
                && _.isEqual(table.table_schema, queryTable.table_schema),
            );

            const id = `database-table-${index}`;

            return table.table_schema === props.selectedSchema
              && filterTable(table, props.searchExpr)
              && (
                <DatabaseTable
                  data={constructData(table)}
                  checked={checked}
                  key={id}
                  id={id}
                />
              );
          })}
        </div>
      </Scrollbars>
    </div>
  );
};

DatabaseViewer.propTypes = {
  data: PropTypes.shape({ table_type: PropTypes.string, table_name: PropTypes.string }),
  constraints: PropTypes.arrayOf(PropTypes.shape({})),
  columns: PropTypes.arrayOf(PropTypes.shape({})),
  tables: PropTypes.arrayOf(PropTypes.shape({})),
  queryTable: PropTypes.arrayOf(PropTypes.shape({})),
  selectedSchema: PropTypes.string,
  searchExpr: PropTypes.string,
};

const mapStateToProps = store => ({
  tables: store.database.tables,
  schemas: store.database.schemas,
  selectedSchema: store.database.selectedSchema,
  constraints: store.database.constraints,
  columns: store.database.columns,
  queryTable: store.query.tables,
  searchExpr: store.database.searchExpr,
});

export default connect(mapStateToProps)(DatabaseViewer);
