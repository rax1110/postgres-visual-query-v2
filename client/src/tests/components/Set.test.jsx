import React from 'react';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import { Draggable } from 'react-beautiful-dnd';
import { CustomInput } from 'reactstrap';
import { Set } from '../../components/Set';

describe('Component: Set', () => {
  let props;
  let mockStore;
  let component;

  beforeEach(() => {
    props = {
      language: { code: 'eng' },
      set: {
        id: 0,
        subqueryId: 1,
      },
      activeQuery: {
        id: 1,
        queryName: 'Query 1',
      },
      queries: [
        {
          id: 1,
          queryName: 'Query 1',
        },
        {
          id: 2,
          queryName: 'Query 2',
        },
      ],
      updateSet: jest.fn(),
      removeSet: jest.fn(),
    };

    mockStore = configureMockStore([thunk]);
  });

  it('renders correctly', () => {
    component = shallow(<Set {...props} store={mockStore()} />);

    expect(component).toMatchSnapshot();
  });

  test('Set handleTypeChange calls updateSet once', () => {
    component = shallow(<Set {...props} store={mockStore()} />);

    const componentBody = shallow(
      component.find(Draggable).prop('children')({}),
    );

    componentBody
      .find(CustomInput)
      .first()
      .simulate('change', {
        target: { value: 'union' },
        preventDefault: jest.fn(),
      });

    expect(props.updateSet).toBeCalled();
  });

  test('Set handleQueryChange calls updateSet once', () => {
    component = shallow(<Set {...props} store={mockStore()} />);

    const componentBody = shallow(
      component.find(Draggable).prop('children')({}),
    );

    componentBody
      .find(CustomInput)
      .at(1)
      .simulate('change', {
        target: { value: 1 },
        preventDefault: jest.fn(),
      });

    expect(props.updateSet).toBeCalled();
  });

  test('Set handleRemove calls removeSet once', () => {
    component = shallow(<Set {...props} store={mockStore()} />);

    const componentBody = shallow(
      component.find(Draggable).prop('children')({}),
    );

    componentBody
      .find('Button')
      .first()
      .simulate('click');

    expect(props.removeSet).toBeCalled();
  });
});
