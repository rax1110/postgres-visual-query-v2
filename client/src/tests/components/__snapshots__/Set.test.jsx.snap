// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Set renders correctly 1`] = `ShallowWrapper {}`;
